File: Tree/BST.py
# Practice 5. Binary Search Tree
import sys
sys.setrecursionlimit(10**7)
from collections import deque
BUILD = 'B'
FIND_MIN = 'm'
FIND_MAX = 'M'

# Node implementation
class TreeNode:
  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None

class BinarySearchTree:
  def __init__(self):
    self.root = None

  # Return True if tree is empty; False otherwise
  def isEmpty(self):
    return (self.root == None)

  def findPlace(self, data):
    ptr = self.root
    if ptr.data < data:
      if ptr.left == None:
        return ptr
      else:
        return self.findPlace(ptr.left)
    elif ptr.data > data:
      if ptr.right == None:
        return ptr
      else:
        return self.findPlace(ptr.right)
    else:
      return None
    
  # Given a sequence arr of integers, start index l, the end index r, 
  # build a binary search (sub)tree that contains keys in arr[l], ..., arr[r].
  # Return the root node of the tree
  def arrayToBST(self, arr):
    if arr != sorted(arr) or not len(arr):
      return None
    
    middle = round(len(arr)/2 - 1)
    root = TreeNode(arr[middle])
    if len(arr) == 2:
      root.right = TreeNode(arr[1])
    else:
      root.left = self.arrayToBST(arr[:middle])
      root.right = self.arrayToBST(arr[middle+1:])

    return root

  # Return the node with the minimum value 
  def findMin(self):
    ptr = self.root
    while ptr.left != None:
      ptr = ptr.left
    return ptr

  # Return the node with the maximum value 
  def findMax(self):
    ptr = self.root
    while ptr.right != None:
      ptr = ptr.right
    return ptr

  def _getHeight(self, curr):
    if not curr:
      return 0
    return 1 + max(self._getHeight(curr.left), self._getHeight(curr.right))

  def _printSpaces(self, n, curr):
    for _ in range(int(n)):
      print(" ", end="")
    if not curr:
      print(" ", end="")
    else:
      print(str(curr.data), end="")

  def printTree(self):
    if not self.root:
      return 
    q = deque()
    q.append(self.root)
    temp = deque()
    cnt = 0
    height = self._getHeight(self.root) - 1
    nMaxNodes = 2**(height + 1) - 1
    while cnt <= height:
      curr = q.popleft()
      if len(temp) == 0:
        self._printSpaces(nMaxNodes / (2**(cnt+1)), curr)
      else:
        self._printSpaces(nMaxNodes / (2**cnt), curr)
      if not curr:
        temp.append(None)
        temp.append(None)
      else:
        temp.append(curr.left)
        temp.append(curr.right)
      if len(q) == 0:
        print("\n")
        q = temp
        temp = deque()
        cnt += 1

if __name__ == "__main__":
  tree = BinarySearchTree()
  with open('input.txt', 'r') as inFile:
    lines = inFile.readlines()
  with open('output.txt', 'w') as outFile:
    for line in lines:
      words = line.split()
      op = words[0]
      if op == BUILD:
        data = [int(s) for s in words[1:]]
        tree.root = tree.arrayToBST(data)
        if tree.root:
          outFile.write(BUILD + "\n")
          tree.printTree()
        else:
          raise Exception("BUILD: invalid input")
      elif op == FIND_MIN:
        found = tree.findMin()
        if not found:
          raise Exception("FindMin failed")
        else:
          outFile.write(str(found.data) + "\n")
      elif op == FIND_MAX:
        found = tree.findMax()
        if not found:
          raise Exception("FindMax failed")
        else:
          outFile.write(str(found.data) + "\n")
      else:
        raise Exception("Undefined operator")
        

File: Tree/BST.cpp
// Practice 5. Binary Search Tree
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <queue>
using namespace std;
const char BUILD = 'B';
const char FIND_MIN = 'm';
const char FIND_MAX = 'M';

class TreeNode {
public:
  int key;
  TreeNode* left;
  TreeNode* right;
  TreeNode(int data) {
    key = data;
    left = nullptr;
    right = nullptr;
  }
};

class BinarySearchTree {
public:
  TreeNode* root;
  BinarySearchTree(): root(nullptr) {}

  TreeNode* arrayToBST(vector<int>&);
  TreeNode* findMin();
  TreeNode* findMax();
  void printTree();
private:
  int _getHeight(TreeNode*);
  void _printSpaces(double, TreeNode*);
};

// Given a sequence arr of integers, start index l, the end index r, 
// build a binary search (sub)tree that contains keys in arr[l], ..., arr[r].
// Return the root node of the tree
TreeNode* BinarySearchTree::arrayToBST(vector<int>& arr) {
  if (arr.size() == 0) return nullptr;
  for (int i=0; i<arr.size()-1; i++)
    if (arr[i] > arr[i+1]) return nullptr;
  
  int middle = ceil((float)arr.size() / 2) - 1;
  TreeNode* root = new TreeNode(arr[middle]);

  if (arr.size() == 2)
    root->right = new TreeNode(arr[1]);
  else {
    vector<int> left_item(arr.begin(), arr.begin() + middle);
    vector<int> right_item(arr.begin() + middle + 1, arr.end());
    root->left = BinarySearchTree::arrayToBST(left_item);
    root->right = BinarySearchTree::arrayToBST(right_item);
  }

  return root;
}

// Return the node with the minimum value 
TreeNode* BinarySearchTree::findMin() {
  TreeNode* ptr = root;
  while (ptr->left != nullptr)
    ptr = ptr->left;
  return ptr;
}

// Return the node with the maximum value 
TreeNode* BinarySearchTree::findMax() {
  TreeNode* ptr = root;
  while (ptr->right != nullptr)
    ptr = ptr->right;
  return ptr;
}

int BinarySearchTree::_getHeight(TreeNode* curr) {
  if (curr == nullptr) 
    return 0;
  return 1 + max(_getHeight(curr->left), _getHeight(curr->right));
}

void BinarySearchTree::_printSpaces(double n, TreeNode* curr) {
  for(double i = 0; i < n; ++i) 
    cout<<" ";
  if (curr == nullptr)
    cout<<" ";
  else
    cout<<curr->key;
}

void BinarySearchTree::printTree() {
  if (root == nullptr)
    return;
  queue<TreeNode*> q;
  q.push(root);
  queue<TreeNode*> temp;
  int cnt = 0;
  int height = _getHeight(root) - 1;
  double nMaxNodes = pow(2, height + 1) - 1;
  while (cnt <= height) {
    TreeNode* curr = q.front();
    q.pop();
    if (temp.empty())
      _printSpaces(nMaxNodes / pow(2, cnt + 1) + height - cnt, curr);
    else
      _printSpaces(nMaxNodes / pow(2, cnt), curr);
    if (curr == nullptr) {
      temp.push(nullptr);
      temp.push(nullptr);
    }
    else {
      temp.push(curr->left);
      temp.push(curr->right);
    }
    if (q.empty()) {
      cout<<endl<<endl;
      q = temp;
      queue<TreeNode*> empty;
      swap(temp, empty);
      ++cnt;
    }
  }
}

int main() {
  BinarySearchTree tree;
  ifstream inFile("input.txt");
  ofstream outFile("output.txt");
  string line;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(1));
    int k;
    TreeNode* found = nullptr;
    vector<int> data;
    switch(op)
    {
      case BUILD:
        while (iss >> k)
          data.push_back(k);
        tree.root = tree.arrayToBST(data);
        if (tree.root) {
          outFile<<BUILD<<endl;
        }
        else {
          cerr<<"BUILD: invalid input"<<endl;
          exit(1);
        }
        tree.printTree();
        break;
      case FIND_MIN:
        found = tree.findMin();
        if (found == nullptr) {
          cerr<<"FindMin failed"<<endl;
          exit(1);
        }
        else {
          outFile<<found->key<<endl;
        }
        break;
      case FIND_MAX:
        found = tree.findMax();
        if (found == nullptr) {
          cerr<<"FindMax failed"<<endl;
          exit(1);
        }
        else {
          outFile<<found->key<<endl;
        }
        break;
      default:
        cerr<<"Undefined operator"<<endl;
        exit(1);
    }
  }
  outFile.close();
  inFile.close();
}
